### DISCUSSION TITLE
DOM context with shortcuts

### DISCUSSION THREAD

**Discussion by yankeeinlondon** [unocss/unocss#916] [Q&A]
Is there any way to gain "surrounding context" on a class token when processing shortcuts and variants? 

So, for instance, I might have variant called `lg` which I use to control the size/scaling of several components. Unlike the popular tailwind/windi `lg:` variant which behaves by creating media queries for different viewport sizes, I'm looking for a uniform way to express size that _varies_ by where it is applied. To aid this "for instance", imagine there are two components: `abc` and `xyz` and both components have meaningful use of this new "lg" variant but for `abc` it means changing the text/font size whereas for `xyz` it is mainly about padding, margin, and border radius. How can I construct a combination of shortcut and variant where the DOM element being operated on can be interrogated? 

Note, this is one of two issue areas which i'm exploring for incorporating Unocss into library code versus userland code. The second issue also revolves around a "context" issue but is more demanding. There are many cases where it would be useful to create variant which is applied to the root node of a component but the _scope_ of the styling is then moved into the interior of the component. For this I'd use the example of an input element with a label. Let's say i have a component called `my-input` and I want to make the label red but the input text blue. What I'd like to do is something like:

```html
<my-input class="label:text-red-500 input:text-blue-500" />
```

And then -- during the _build stage_ -- I'd like to have the `label` variant move the styling to be placed on the label DOM element and similarly the `input` variant placed on the input element. This allows two styles that -- if placed on the wrapping element of the component -- would be conflicting co-exist but it also opens up a lot of useful scenarios for expressing styles in a structured component that are very powerful.

My suspicion is that my first quandry/example either _is_ solvable now or could be made solvable without too much effort. The second one -- which requires modification of the DOM tree (versus just read access to some properties) might be more complicated but eager to hear from others who are further down the rabbit hole as I'm just getting started here.

**Comment by antfu** [comment]
Here is something you might be interested in
https://github.com/nuxt/ui/tree/main/packages/ui

**Reply by yankeeinlondon** [reply]
Will have a look right now

**Reply by yankeeinlondon** [reply]
Ok let me summarize what I see and how I'd ideally like to go a bit further. 

1. **Top Level Component Scoping**

    It seems you're using class _name spacing_ (which could be defined as a prefix or rule) to help facilitate scoping. What I see is:

    1. if the consumer wanted to effect all representations of **NSwitch** you could modify the preset/rules you wanted and the change would be scoped to a particular component type (for instance `n-switch-base`). 
    2. of course, if the consumer introduced css on the root of the component, their "layer" would be given higher priority and allow them greater granularity in over-riding certain aspects of the default style but not all
    3. if the consumer wanted to leverage a rule like `n-transition` they can do this too but as it's scope is across the component library (versus just for the given component) a change to the preset/rule would be global to the library and not just the component but it does open up reusable styles across components. 

    This is decent coverage for top level styling but doesn't addressed interior scope (covered in next section) and has some weaknesses too:

    1. There is too much knowledge a consumer needs to have on whether a given style is scoped locally, globally, with a rule or a preset. They will need to gain an understanding of this if they want to override the defaults for the component. What I was hoping for instead was for the consumer to only really need to know about what rules/presets a component uses and what variants are available.
    2. You can only change between a global setting or very specific CSS properties. It would be nice if components were to expose variants that were particular to themselves.

    I will give examples of this but the examples will become clearer after we discuss the next section ...

2. **Interior Scope**

    A component almost always represents a structured DOM tree rather than a single DOM node. Slots let us provide openings for consumers to make large changes to how some of the interior scope should be styled but it is desirable to have a good "default" for this structure which the user can use in an _inline_ manner without resorting to slot use.

    Sticking with the **NSwitch** example, you have a presets called `n-switch-slider` and `n-switch-thumb` which are placed in the interior of the component and because they don't have slot alternatives you force consumers who want to target that _part_ of the component to use traditional CSS class which could target the `n-switch-slider` class but many consumers today would want to use a utility framework like Tailwind or Windi which unfortunately has no means to "reach into the component" and put the utility class into the right place.

    So, here's my idealized "for instance", to make this more dramatic I'm assuming that a switch not only has a master label but can have value labels for left/right (currently you're only representing binary values but it's common to be able to express the mean of the two states as properties):

    ```vue
    <n-switch class="large left:text-green-500 right:text-red-500 slider:bg-blue-500" left="yes" right="no" />
    ```

    In this example you'd have the following behavior:
    - The `large` class name would be scoped to `n-switch:large` and would apply the appropriate styling to not just base of the component but also be able to adjust shadow sizes, rounding, etc. across various inner scopes in the component
    - The `left` variant would be defined to be _moved_ to the left label describing that state and once there it would be _unwrapped_ to just be `text-green-500`; same thing with the `right` and `slider` variants.
    - the final step -- though not illustrated here -- is that if you wanted as a library author, it might be nice to have a default scope for various presets in the component. For instance, in a text input which has both a label property and the input text, it might be nice to let a consumers expression of `text-red-500` as being associated to the `input`'s text and _not_ the `label`'s text. The consumer could still explicitly target the label's text color with `label:text-red-500`

I actually have a library that does all of this already and it's a really nice consumer experience but there is too much happening at run-time and I'd really prefer to work off of UnoCSS as a base.

**Reply by yankeeinlondon** [reply]
@antfu are there aspects I missed in your example or any pointers you could provide on how to achieve the goals above

**Reply by antfu** [reply]
While I understand the goal you want to achieve, I generally think context-specific utils actually obey the idea of being "atomic".

It's might be possible to achieve your example with a custom extractor, however, I think it will have quite a lot hard caveats:

```html
<!-- static, fine -->
<n-switch class="large left:text-green-500 right:text-red-500 slider:bg-blue-500" />
```

```vue
<script setup>
defineProps<{
  customClass?: string // probably "slider:bg-blue-500", or can be anything
}>()

let myClass = computed(() => 'left:text-green-500')
</script>

<template>
  <n-switch 
    class="large"
    :class="[myClass, customClass]"
  />
</template>
```

In this case, how do you know `left:text-green-500` is associated with `n-swtich` **at build time**?

**Reply by yankeeinlondon** [reply]
I think this is, however, a characteristic of component design. Consumers would like to treat the "component" as an atomic unit rather than have to try to decompose it's structure in order to interact with it. WRT to scoping of the `left:` variant what I originally was thinking was that when the rule/variants are applied that a reference to the DOM element is made available. This way one can check the component name by looking at the `tagName`. If this DOM reference were read-only then this would allow some scoping scenarios but ideally you could update the DOM reference to _move_ the interior scope into the correct place in the component.

**Reply by yankeeinlondon** [reply]
I'm not sure this will all be entirely intelligible but here's how I do this today ... I have a sidecar TS file for each component which defines styles. Each component is made up of "regions" that are defined like this:

```ts
const SWITCH = defineStyle("switch", (p: P) => {
  const sizeLookup = createLookup<SwitchSizes, Number[]>(
    {
      small: [10, 22],
      normal: [12, 26],
      large: [18, 38],
    },
    () => [12, 26]
  );

  return {
    defaults: (c) =>
      c //
        .bg("bg-gray-300")
        .layout("flex", "flex-row", "items-center", "relative")
        .animation("transition-all", "duration-700", "ease-in-out")
        .effects(isTrue(p.disabled) ? "cursor-default" : "cursor-pointer")
        .effects(shadowLookup(p.shadow, p.size))
        .other(`w-${sizeLookup(p.size)[0]}`, `h-${sizeLookup(p.size)[1]}px`)
        .border(
          "rounded-full",
          "border-1",
          "border-transparent",
          v.focus,
          "ring-2",
          "ring-offset-0",
          "ring-inset",
          "outline-none",
          `ring-${p.color}-500`,
          "border-1",
          "border-transparent"
        ),

    delegate: (r) => [
      r.sizing.width,
      r.sizing.height,
      r.bg.bgColor,
      r.border.ringColor,
      r.border.ringOffsetWidth,
      r.border.ringWidth,
      r.border.borderColor,
      r.border.borderWidth,
    ],
    rules: (r) => [r.layout.display, r.layout.flexDirection, r.spacing.mt, r.spacing.p],
  };
```

**Reply by yankeeinlondon** [reply]
In the above example, if a user uses a `width` or `height` rule at the base component level it will be moved/delegated to the "switch" region.
